app/planets/list/gateway/swapi_gateway.go

```diff
@@ -0,0 +1,48 @@
+package gateway
+
+import (
+	"net/http"
+	"starwars-api-go/app/commons"
+)
+
+type SWAPIGateway interface {
+	GetPlanets(filter Filter) (*http.Response, error)
+}
+
+type swapiGateway struct {
+	client  *http.Client
+	baseURL string
+}
+
+func NewSWAPIGateway() SWAPIGateway {
```

Comment:
Ao invés de retornar a interface, é uma boa retornar o tipo concreto e deixar a definição da interface do lade de quem vai consumir, pra reduzir o acoplamento. Isso vale para os outros locais onde uma interface tá sendo retornada também, Service por exemplo.

Não tem problema em retornar o `swapiGateway` privado no construtor. O que isso ta fazendo é somente impedir que outro pacote crie um `swapiGateway` sem usar o construtor. 

https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b

Reply:
Simplesmente genial esse artigo.